storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala

```diff
@@ -57,66 +57,73 @@ object MergeableStore {
     val (collectedMGetResult, getFailures) = collectWithFailures(mGetResult)
 
     val mPutResultFut: Future[Map[K, Future[Option[V]]]] =
-      collectedMGetResult.map { pairs: Seq[(K, (Option[V], Option[V]))] =>
+      collectedMGetResult.map { pairs: Array[(K, (Option[V], Option[V]))] =>
         val pairMap = pairs.toMap
         val mPutResult: Map[K, Future[Unit]] = store.multiPut(pairMap.mapValues(_._2))
         mPutResult.map { case (k, funit) =>
           (k, funit.map { _ => pairMap(k)._1 })
         }
       }
 
+    val missingFn: K => Future[Option[V]] = { k =>
+      getFailures.flatMap { failures =>
+        Future.exception(failures.getOrElse(k, new MissingValueException[K](k)))
+      }
+    }
+
     /**
-     * Combine original keys with result after put.
+     * Combine original keys with result after put and errors with get.
      */
-    FutureOps.liftFutureValues(keySet, mPutResultFut, getFailures)
+    FutureOps.liftFutureValues(keySet, mPutResultFut, missingFn)
   }
 
   /**
    * Collects keyed futures, partitioning out the failures.
    */
-  private[algebra] def collectWithFailures[K, V](fs:Map[K, Future[V]]): (Future[Seq[(K, V)]], Map[K, Future[Nothing]]) = {
+  private[algebra] def collectWithFailures[K, V](fs: Map[K, Future[V]]): (Future[Array[(K, V)]], Future[Map[K, Throwable]]) = {
```

Comment:
looking at this again it seems like rather than `Either[Throwable, V]` we could just use `Try[V]` which is basically the same.

Then this method signature could be:
`def collectWithFailures[K, V](fs: Map[K, Future[V]]): Future[Map[K, Try[V]]]` which actually seems useful enough to have even in twitter util.

In any case, the first thing you do when you get the result is call `.toMap` on the array, so returning the `Map` seems to make sense.

Also, both of these futures are satisfied at the same time, right? So, even if you don't merge the data structures, you could do: `Future[(Map[K, V], Map[K, Throwable])]` so it is clear that all the data comes in at one time.